### ISSUE TITLE
[TD-104] Unrecognized taskwarrior file format

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskserver#104]
_Jeremy John Reeder on 2015-09-09T20:23:24Z says:_ 

 When running `task synchronize` in Taskwarrior 2.4.5 or 2.5.0.beta1, I commonly encounter the following error:
{code}
Sync failed.  The Taskserver returned error: 500 Unrecognized taskwarrior file format.
{code}

Each time I encounter this error, I restore my task database from a backup and go back to Taskwarrior 2.4.4 for a while. Taskserver never has this problem with that version of Taskwarrior.

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2015-09-09T20:23:24Z
Modified: 2017-08-10T17:35:53Z
```

**Comment by taskwarrior** [comment]
_Jeremy John Reeder on 2015-09-09T20:33:33Z says:_ 

 Each time I try out Taskwarrior 2.4.5 or 2.5.0, I do several modifications and several synchronizations over a period of a day or more before I encounter this problem. So it seems that Taskwarrior 2.4.4 does not create a database that's incompatible with Taskwarrior 2.4.5+ or Taskserver 1.1.0, but that Taskwarrior 2.4.5+ does occasionally create these incompatibilities.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2015-09-09T23:18:44Z says:_ 

 This might be an incompatibility with legacy duration values.  Would you run these 2.5.0 commands please, and report the results:

```
$ task _unique recur
$ task udas
```

If any of your UDAs above have type `duration`, please provide a unique list of those values:

```
$ task _unique NAME
```

**Comment by taskwarrior** [comment]
_Jeremy John Reeder on 2015-09-10T03:00:57Z says:_ 

 Well, I do have a duration field. Are those no longer supported, or are they just supported differently than in the past? Anyway, here's the output that you requested.
{code}
> src/task --version
2.5.0.beta1
> src/task _unique recur
1year
3d
5mo
biweekly
monthly
quarterly
weekly
> src/task udas

Name     Type     Label    Allowed Values Default Usage Count
distance numeric  Distance                        20
duration duration Duration                        44
location string   Location                        106
priority string   Priority H,M,L,                 746
proxy    string   Proxy                           8

5 UDAs defined
> src/task _unique duration
1.5h
1800
1h
2h
3.5h
30min
34min
35min
5400
6300
PT23M15S
PT30M7S
PT34M59S
PT34M5S
PT37M37S
PT9M21S
PT9M3S
> 
{code}

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2015-09-10T03:44:44Z says:_ 

 Thanks for promptly supplying data.

Duration fields are still supported, no change there.  But I did fix a bug where a zero-length duration was being stored as `P0S` which is wrong, and should be `PT0S`.  However, that doesn't seem to be the problem here.

I suspect those `1800`, `5400` and `6300` values are to blame.  I suspect if those were changed to their corresponding `PT30M`, `PT1H30M` and `PT1H45M` values, the problem goes away.  This would mean I failed to add legacy value conversions.

**Comment by taskwarrior** [comment]
_Jeremy John Reeder on 2015-09-10T12:11:12Z says:_ 

 Okay. I converted those values with the following commands. I'm switching back to version 2.5.0.beta1, and we'll see if that fixed it.
{code}
~/.task> sed -ir 's/duration:"1800"/duration:"PT30M"/' *.data
~/.task> sed -ir 's/duration:"5400"/duration:"PT1H30M"/' *.data
~/.task> sed -ir 's/duration:"6300"/duration:"PT1H45M"/' *.data
{code}

**Comment by taskwarrior** [comment]
_Tomas Babej on 2015-09-10T12:40:56Z says:_ 

 Thanks for raising the issue, let us know.

**Comment by taskwarrior** [comment]
_Jeremy John Reeder on 2015-09-11T16:55:35Z says:_ 

 It's broken again.
{code}
> task sync
Sync failed.  The Taskserver returned error: 500 Unrecognized taskwarrior file format.
> task _unique recur
1year
3d
5mo
biweekly
monthly
quarterly
weekly
> task _unique duration
1.5h
1h
2h
3.5h
30min
34min
35min
PT1H30M
PT1H45M
PT23M15S
PT30M
PT30M7S
PT34M59S
PT34M5S
PT37M37S
PT9M21S
PT9M3S
{code}

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2015-09-11T16:58:38Z says:_ 

 So much for my little theory.  Okay, thank you, time to investigate deeper...

**Comment by taskwarrior** [comment]
_John Murphy on 2015-11-08T19:52:51Z says:_ 

 I've been having this issue fairly regularly, too.  To "fix" it I edit my backlog.data to just contain the UUID (and remove any other JSON there).  The sync appears to have "worked", despite the error.

However, I have noticed that many tasks with dependencies are getting inappropriately escaped in the "dependency" attribute, eg.:

{code}
[depends:"&open;\"d138ef16-efe8-409e-b633-4e05600b34fe\"&close;" description:"..." entry:"1445479807" modified:"1445479807" status:"pending" tags:"..." uuid:"784c5223-e06a-49a7-a993-5c9f2f8941c3"]
{code}

This causes those tasks to not be editable via "task edit" until the dependencies are cleared.  It doesn't seem to happen all the time, and at first I thought it was due to using a mix of 2.4 and 2.5 tw clients, but I'm pretty sure I've got everything on 2.5 now and I still see this issue from time to time.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2015-11-08T20:15:14Z says:_ 

 Great, thanks.  Bear in mind that if you delete the JSON in backlog.data, those lines represented local tasks that were not yet synced, so you may now have unsynced data.  To remedy this, you have two options that come to mind:

1. Assuming your client is up to date, run `task sync init` and it will upload all pending tasks.  Then your other clients should be fine with just a `task sync`.

2. Update every task (`task status:pending modify +useless_tag; task +useless_tag modify -useless_tag`), and then `task sync`.

The `&close;` encodings is a separate bug, which (needless to say) should not occur.  But it did, therefore bug.

**Comment by taskwarrior** [comment]
_John Murphy on 2015-12-20T15:18:04Z says:_ 

 OK, I think I can reliably reproduce this (at least of the scenario affecting me).

With two 2.5.0 clients, both starting in a fully synced state:

On client 1:

{code}
[jackie:/home/jackie] master ± task add foo
Created task 379.
[jackie:/home/jackie] master ± task add bar dep:379
Created task 380.
[jackie:/home/jackie] master ± task sync
Syncing with REDACTED

Sync successful.  2 changes uploaded.
{code}

On client 2:

{code}
[jackie:~] osx ± task sync
Syncing with REDACTED
     add da4d1e04-be2a-442b-8f00-c0954160b976 'foo'
     add 451e1922-3fe4-41df-8b3e-d711c0ef1777 'bar'

Sync successful.  2 changes downloaded.
[jackie:~] osx ± task foo

ID  Age  Description Urg
378 1min foo            0

1 task
[jackie:~] osx ± task 378 done
Completed task 378 'foo'.
Unblocked 379 'bar'.
Completed 1 task.
You have more urgent tasks.
[jackie:~] osx ± task sync
Syncing with REDACTED

Sync successful.  1 changes uploaded.
[jackie:~] osx ± task bar
[task next ( bar )]

ID  Age  Description Urg
378 1min bar            0

1 task
[jackie:~] osx ± task 378 done
Completed task 378 'bar'.
Completed 1 task.
You have more urgent tasks.
[jackie:~] osx ± task sync
Syncing with REDACTED

Sync failed.  The Taskserver returned error: 500 Unrecognized taskwarrior file format.
{code}

At this point, the contents of `~/.task/backlog.data` is:

{code}
58ecfb37-c7bb-43f4-a3f8-f9ca52217a30
{"depends":["["da4d1e04-be2a-442b-8f00-c0954160b976"]"],"description":"bar","end":"20151220T150529Z","entry":"20151220T150410Z","modified":"20151220T150529Z","status":"completed","uuid":"451e1922-3fe4-41df-8b3e-d711c0ef1777"}
{code}

EDIT: Cleaning up the `depends` attribute allows the sync to proceed.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2015-12-22T16:31:03Z says:_ 

 Thank you for the reproducible case, that's perfect.  Will fix.

**Comment by taskwarrior** [comment]
_Dirk Deimeke on 2016-02-05T15:04:48Z says:_ 

 I get this errors really often. What about releasing a bug fixed version of Taskserver? Maybe taskd 1.1.1?

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2016-02-05T18:13:17Z says:_ 

 1.2.0 is nearly ready to go.  But we don't have a fix for this yet.

**Comment by taskwarrior** [comment]
_Reg on 2016-02-07T21:20:14Z says:_ 

 I also get this error a lot (same scenario as John Murphy with the dependencies), but I cannot reproduce it anymore since I put json.depends.array=off in the .taskrc of all my clients. All blocking tasks are also highlighted properly since then.

Any reason why this is not the default setting? I'm worried it has side-effects I haven't noticed yet.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2016-02-07T21:25:23Z says:_ 

 Hey, that's great info - thanks for the clue!

**Comment by taskwarrior** [comment]
_Reg on 2016-02-20T06:38:43Z says:_ 

 Since I made the json.depends.array change, I get errors like this from time to time:

$ task
Unrecognized Taskwarrior file format. in ~/.task/pending.data at line 60

The line in question consists looks like this: ^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@ (repeated many, many times).

Deleting it solves the issue. I do not notice any loss of data.

I haven't figured out what's causing it yet. It only happens on one of my devices, always the same device. 

My apologies if this isn't the right place to post this.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2016-02-20T13:04:54Z says:_ 

 Reg can you tell me which Taskwarrior version you are using, with this:

```
$ task diag | egrep 'Built|Commit'
```

**Comment by taskwarrior** [comment]
_Reg on 2016-02-20T16:33:37Z says:_ 

 Interesting, the device on which this occurs (Android/termux) has:
      Built: Oct 26 2015 19:40:31

But my other device has (ArchLinux):
      Built: Dec  7 2015 07:03:45

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2016-02-21T16:27:48Z says:_ 

 The problem is that Taskwarrior 2.5.0 (just the one release) defaulted to `json.depends.array=on`, and all Taskserver versions (1.0.0, 1.1.0) are unaware of this format, the result being unreadable data.

This is resolved, but not in a nice clean way.  It's a two-step process:
*  Taskserver 1.2 can now read array and non-array dependencies, but emits the non-array form.
*  Taskwarrior 2.5.1 can now read array and non-array dependencies, and can emit either, depending on 'json.depends.array'.
*  Taskwarrior 2.5.1 now defaults to 'json.depends.array=off'.

At a future time, when Taskwarrior 2.5.1 and Taskserver 1.2.0 are released, Taskwarrior and Taskserver will be modified to no longer support the non-array form for dependencies. That will leave the software in a state where it reads both the array and non-array forms, but only writes the array form.

**Comment by taskwarrior** [comment]
_Antonio Huete Jimenez on 2017-03-01T18:55:10Z says:_ 

 Hi,

Followed these steps (as recommended by Paul). Client version 2.5.1 and server version 1.1.0:

1) Make a backup of tx.data on the server
2) Delete tx.data on the server
3) Make sure you have json.depends.array=off in your clients' ~./taskrc
4) Backup your client data (~/.task) on the client you're going to use for working around this issue.
5) In that client, run "task sync init"
6) In the other clients you may have remove the following files located in ~/.task/ : pending.data, completed.data, backlog.data

Best regards,
Antonio Huete

**Comment by MartyLake** [comment]
For posterity, I confirm that cleaning the backlog.data from it's broken depends fixes this issue. Here are the lines I appended to my shared bashrc to detect and fix the problem faster next time:

```

if [[ "$(task --version)" == "2.5.0" ]]; then
    echo "bugged version of taskwarrior 2.5.0, please upgrade or the taskfile will get corrupt :/"
fi

if [[ $(cat ~/.task/backlog.data | grep 'depends":\[:\[') ]]; then
    echo "corrupted task/backlog.data, remove array of depends"
fi
```

**Comment by corbolais** [comment]
Well, well. There it strikes again. But the cause and solution are different this time:

**TL;DR: see at the far end of this post.**

On Debian client:
```
Syncing with srv:53589
Sync failed.  The Taskserver returned error: 500 Unrecognized taskwarrior file format.
```

Two clients (one Ubuntu, one Debian), one taskd srv.

- Client 1: Debian buster/sid x86 64bit
  - task 2.5.1 built for linux
  - dpkg: taskwarrior    2.5.1+dfsg-6 amd64
  - task diag:
```
task 2.5.1
   Platform: Linux

Compiler
    Version: 7.2.0
       Caps: +stdc +stdc_hosted +LP64 +c8 +i32 +l64 +vp64 +time_t64
 Compliance: C++11

Build Features
      CMake: 3.9.5
    libuuid: libuuid + uuid_unparse_lower
  libgnutls: 3.5.16
 Build type: None
```

- Client 2: Ubuntu Cosmic x86 64bit
  - task 2.5.1 built for linux
  - dpkg: taskwarrior    2.5.1+dfsg-6 amd64
  - task diag:
```
task 2.5.1
   Platform: Linux

Compiler
    Version: 7.2.0
       Caps: +stdc +stdc_hosted +LP64 +c8 +i32 +l64 +vp64 +time_t64
 Compliance: C++11

Build Features
      CMake: 3.9.5
    libuuid: libuuid + uuid_unparse_lower
  libgnutls: 3.5.8
 Build type: None
```

- Server: Debian buster/sid ARMv7
  - taskd          1.1.0+dfsg-3 armhf
  - Empty: /var/log/taskd.log
  - taskd 1.1.0 built for linux
  - this is the latest version available in the pkg repos.

Note: libgnutls Ubuntu Cosmic version is older than the Debian version.

Sync from the Ubuntu client is OK.

Sync from Debian client yields "`Sync failed.  The Taskserver returned error: 500 Unrecognized taskwarrior file format.`"


**So, here's the real deal:** 

The **task** entry in question **does not feature a dependency**.

The clients `.taskrc` does not contain a json.* directive. The `taskrc` man page tells "`json.depends.array=on [...] Defaults to on.`" This is in contradiction to the above post stating: "`Taskwarrior 2.5.1 now defaults to 'json.depends.array=off'.`"

```
Mar  1 11:30:01 srv taskd[17485]: 2019-03-01 10:30:01 [11] patch modify modified=1551435284
Mar  1 11:13:45 srv taskd[17485]: 2019-03-01 10:13:45 [2] Merge result {"description":"foo ","due":"20190116T230000Z","entry":"20181023T193806Z","mask":"+-----++++++++-----+++++++++++++------------- ,"modified":"20190301T100001Z","priority":"H","project":"test.again","recur":"daily","status":"recurring","tags":["IT","taskd"],"uuid":"edee920c-cb8c-47e7-855b-4f611883460d","annotations":[{"entry":"20181023T212628Z","description":"https:\/\/www.w3.org\/"}]}
Mar  1 11:30:01 srv taskd[17485]: 2019-03-01 10:30:01 [11] Unrecognized taskwarrior file format.
Mar  1 11:30:01 srv taskd[17485]: 2019-03-01 10:30:01 [11] Serviced in 0.279997s
```
Take note of the **annotation including a URL description**. The json array above is shorter for demonstration. Really, it contains about 10+ annotations with URL descriptions. However, removing any annotation yields:
```
Mar  1 12:20:01 srv taskd[17485]: 2019-03-01 11:20:01 [33] Merge result {"description":"foo","due":"20190116T230000Z","entry":"20181023T193806Z","mask":"+-----++++++++-----+++++++++++++-------------","modified":"20190301T111952Z","priority":"H","project":"test.again","recur":"daily","status":"recurring","tags":["IT","taskd"],"uuid":"edee920c-cb8c-47e7-855b-4f611883460d","annotations":[]}
Mar  1 12:20:01 srv taskd[17485]: 2019-03-01 11:20:01 [33] Unrecognized taskwarrior file format.
Mar  1 12:20:01 srv taskd[17485]: 2019-03-01 11:20:01 [33] Serviced in 0.319996s
```
Uhm. Not what I was hoping for. What now? Editing `.taskrc` to include `json.depends.array=off`. Result? The same as above.

What does not work either:
- deleting this single entry, none of the other recurring tasks.
- deleting all recurring tasks including parent (in fact "uuid":"edee920c-cb8c-47e7-855b-4f611883460d" _was_ the parent)

Question aside: is the "Mask" tag growing indefinitely in size?

Any ideas? This is a huge blocker now. I am unable sync the Debian client _at all_ which is a great pain ITA.

**Comment by sdondley** [comment]
I'm getting this now, too.

I upgraded to 1.2 recently. Clients running 2.6.1.

I have one client working with the server just fine. But when I `try task sync` init on another machine, I get the error: "Unrecognized Taskwarrior file format or blank line in data."

**Comment by pritamdodeja** [comment]
This might be useful to someone so putting it out there.  On taskwarrior for Android, if you create a task with just one word, and there's a trailing space after that task, it corrupts the data and you'll get the same message as this thread.  In order to recover from this issue, I used a hammer.  The hammer I used was as follows:

1. Connect via adb to the android device: ```adb -s <serialnumber> shell```
2. Find where your taskwarrior data is ```find . -name completed.data 2> /dev/null```
3. Go into that directory ```cd /storage/emulated/0/Android/data/kvj.taskw/files/<id>/data/```
4. It's hammer time (deletes all timewarrior data on device, assumption here is that you have a taskd server or a backup somewhere) ```rm *.data```
5. Sync using the taskwarrior for Android UI

Taskwarrior is amazing.  Thank for for writing this software!

**Comment by btwe** [comment]
> I used a hammer.

A huge one. You can usually access the filesystem on android with any file manager (in most cases). Like the stock file manager, total commander, termux, ... but yeah adb is there too, but would need to have another PC

I also use kvj.taskw .. but well this is ancient software. Last commit 2016-03. So I am currently slowly migrating to termux (they provide taskwarrior in v 2.6.2 currently). And if you are living in the terminal world like me, you'll find it beautiful too. I also got a small keyboard now and on the go I have a nice feeling of having a proper console at hand :+1:

**Comment by pritamdodeja** [comment]
> 

I thought with the latest android versions they've disabled access to the (at least) these files through the Files application.  I also prefer living in terminal land, but when you're driving and a thought comes, it's much easier to open up this app and use the microphone keyboard to capture it.  I've played around a little with termux on a chromecast, although I have been experimenting with mosh + tmux for terminal access.
